### DISCUSSION TITLE
improving readability of clicked documents

### DISCUSSION THREAD

**Discussion by thisisnottaken** [ImageMagick/ImageMagick#2718] [Digital Image Processing]
hi,
You must be aware of camscanner android app using which you can enhance the readability of a document image.
It is a filter which makes greyish clicked image far better with white background and sharp text.
This transform is useful before printing the clicked image.
I think it is more than histogram equalization .
Is there any such tool in image magick ?

Thanks.

**Comment by snibgo** [comment]
Perhaps "-contrast-stretch" and a sharpen.

If you show a before and after example, perhaps we can advise.

**Reply by thisisnottaken** [reply]
ok i will post original and cam scanner images soon .

**Comment by thisisnottaken** [comment]
Ok .
Sorry for delay .
Here are the images - 
(1)original - https://ibb.co/GxNdcwm
(2)medium filter - https://ibb.co/zH0h8mB
(3)extreme filter - https://ibb.co/nCdRFVm

Medium filter is most suitable for taking printout as it keeps face and contrasts the text with background.
Can you suggest options to achieve this or get better results using image magick ?

Thanks

**Comment by snibgo** [comment]
This works well:
```
magick one-orig.jpg -contrast-stretch 1x50% x.png
```

**Comment by thisisnottaken** [comment]
yeah , that works well .
How contrast stretch works and why you chose 1x50% ?

**Comment by fmw42** [comment]
You can do local area thresholding for a mask and then use the mask to replace the gray background with white and then do some sharpening.

Or in place of local area thresholding, you can do division normalization by dividing a blurred copy of the image by the original image.

See -lat for local area thresholding.

Also see my textcleaner script at my web site at http://www.fmwconcepts.com/imagemagick

**Reply by thisisnottaken** [reply]
> You can do local area thresholding for a mask and then use the mask to replace the gray background with white and then do some sharpening.
> 
> Or in place of local area thresholding, you can do division normalization by dividing a blurred copy of the image by the original image.
> 
> See -lat for local area thresholding.
> 
> Also see my textcleaner script at my web site at http://www.fmwconcepts.com/imagemagick

You are a master in image processing.
I tried your text cleaner script .
It is pretty useful when image has only text.
Here is the output with parameters ' -g -e stretch -f 25 -o 20 -s 1 ' --> https://ibb.co/dLqhTXr
As you can see that except the face , everything is clear .So, your script is best for text image processing .
It seems the face has regions with slight variation in color , so all gets erased except the face outline. Can we avoid (/do less processing) that for the face part of my image in your script ?

 
Thanks.

**Reply by fmw42** [reply]
Not directly. You can create a mask for the region you want to ignore. Then process the whole image with text cleaner. Then use the mask to merge the processed and original images so that the masked white area uses the original image and the black area uses the processed image.

**Comment by snibgo** [comment]
> How contrast stretch works and why you chose 1x50% ?

See https://imagemagick.org/script/command-line-options.php#contrast-stretch

Text documents, with black text on white background, are mostly white. "-contrast-stretch 1x50%" increases contrast so about 1% of pixels are black and 50% are white.

**Reply by thisisnottaken** [reply]
> > How contrast stretch works and why you chose 1x50% ?
> 
> See https://imagemagick.org/script/command-line-options.php#contrast-stretch
> 
> Text documents, with black text on white background, are mostly white. "-contrast-stretch 1x50%" increases contrast so about 1% of pixels are black and 50% are white.

Got it .
Your approach is simple and highly rewarding.
Thanks.

**Comment by sourabhkhemka** [comment]
I don't have any intuition of ImageMagik but one of my recent OpenCV projects discusses the complete image processing steps you could follow to output CamScanner like images. You may find wiki of this [repository](https://github.com/sourabhkhemka/DocumentScanner/)  somewhat helpful in achieving similar results in ImageMagik also

**Reply by thisisnottaken** [reply]
I can try that.
Can you give a simple step by step example of how to use your script.
Let the test case be that I have an image and I want to sharpen it .
https://github.com/sourabhkhemka/DocumentScanner/wiki is not so elaborate.
Thanks.

**Reply by sourabhkhemka** [reply]
I think using DocumentScanner/RScan/scan/*scan.py* file to understand the algorithm's operation will be a simpler way out.

To sharpen the image like magic color mode does in CamScanner, you need two functions namely *blackpointselect* and *whitepointselect*. The former makes all pixels below certain value black and the latter changes all pixels above certain threshold to completely white pixels i-e pixel value is made 255. Using these two you operations one by one on your image you could get the magic color effect ( in the repo it's denoted as *RMode* )

You should refer to [this](https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction)  wiki page also the page named "RMODE: The Magic Color Effect" to understand the workflow. You can also directly look at the python file scan.py. That's a pretty small script.

I tried my best to make the wiki explanatory 😅 but okay, thanks for the feedback 🙂

**Reply by fmw42** [reply]
See -range-threshold or -color-threshold at https://imagemagick.org/script/command-line-options.php

**Reply by thisisnottaken** [reply]
> I think using DocumentScanner/RScan/scan/_scan.py_ file to understand the algorithm's operation will be a simpler way out.
> 
> To sharpen the image like magic color mode does in CamScanner, you need two functions namely _blackpointselect_ and _whitepointselect_. The former makes all pixels below certain value black and the latter changes all pixels above certain threshold to completely white pixels i-e pixel value is made 255. Using these two you operations one by one on your image you could get the magic color effect ( in the repo it's denoted as _RMode_ )
> 
> You should refer to [this](https://github.com/sourabhkhemka/DocumentScanner/wiki/Scan:-Introduction) wiki page also the page named "RMODE: The Magic Color Effect" to understand the workflow. You can also directly look at the python file scan.py. That's a pretty small script.
> 
> I tried my best to make the wiki explanatory 😅 but okay, thanks for the feedback 🙂

okay . 
i will try and get back .
Should I post my queries about your software somewhere else since this is imagemagick forum ?
Thanks.

**Reply by sourabhkhemka** [reply]
Yes sure @thisisnottaken You can post @ [repository's discussion page](https://github.com/sourabhkhemka/DocumentScanner/discussions/1)
